### PR TITLE
Add kcp-config argument wrapper

### PR DESCRIPTION
**Pull request by kylape**
`usage: kcp-config [config file] [cmd] [args...]`

This allows kcp to be configured using a yaml file.  The yaml document needs to be a flat map (i.e. no hierarchy), and each key/value pair will be added as a command-line argument to kcp.

Example config.yaml:

    enable-controllers: true
    listen: localhost:6443

with this invocation:

    kcp-config config.yaml kcp start

will translate to:

    kcp start --enable-controllers=true --listen=localhost:6443

### CODE REVIEW

File: cmd/kcp-config/main.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	"fmt"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+func main() {
+	contents := map[string]interface{}{}
```
Comment (sttts) [inline cmd/kcp-config/main.go:13]: nit: move where used, just above Unmarshal

File: cmd/kcp-config/main.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		fmt.Printf("error: %v\n", err)
+		os.Exit(1)
+	}
+
+	args := os.Args[2:]
+	for k, v := range contents {
+		args = append(args, fmt.Sprintf("--%s=%v", k, v))
+	}
+
+	cmd, err := exec.LookPath(os.Args[2])
```
Comment (stevekuznetsov) [inline cmd/kcp-config/main.go:38]: When can this be anything other than `kcp`? Why is it user-configurable?

File: cmd/kcp-config/main.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+	fname := os.Args[1]
+
+	data, err := ioutil.ReadFile(fname)
+	if err != nil {
+		fmt.Printf("Error reading config file %s: %v\n", fname, err)
+		os.Exit(1)
+	}
+
+	contents := map[string]interface{}{}
```
Comment (stevekuznetsov) [inline cmd/kcp-config/main.go:26]: We're passing it as a set of strings to the command, so why is this not `map[string]string`?

How would this allow passing a flag more than once?

**Review by stevekuznetsov — COMMENTED** [review]
What sort of use-case is this supporting?

**Comment by kylape** [comment]
I'm wanting to supply command line arguments from a `ConfigMap` [based on this conversation](https://kubernetes.slack.com/archives/C021U8WSAFK/p1643924314456659).

File: cmd/kcp-config/main.go (hunk 2 above)
Comment (kylape) [reply cmd/kcp-config/main.go:38]: I'm thinking this wrapper could be used for any of the commands in this project that we'd want to run in a container (and have their args supplied in a `ConfigMap`).

File: cmd/kcp-config/main.go (hunk 3 above)
Comment (kylape) [reply cmd/kcp-config/main.go:26]: The yaml document could have boolean or int types that get converted into strings.  We could probably add some type checking to ensure a flat hierarchy (i.e. no nested maps).

You currently can't pass more than one flag.  If we need that, perhaps we could consider a special delimiter for that?

**Comment by stevekuznetsov** [comment]
I'm ambivalent. I am not sure I understand why passing flags is hard or why we are doing something different than existing k8s apiserver config (stored in etcd?)

**Comment by kylape** [comment]
Passing flags isn't hard per se, but it seems intuitive to me to separate out the configuration of KCP from the more static pod spec.  I'm happy to use flags exclusively if that's the typical way k8s apiservers are managed.  If/when we write an operator to manage kcp, this will likely become a moot point anyways.

> why we are doing something different than existing k8s apiserver config (stored in etcd?)

I'm not sure exactly what you're referring to here.  Perhaps I'm not familiar with how k8s apiservers are normally configured.  Afaict it's just command-line arguments passed in via the container spec.

**Comment by ncdc** [comment]
Discussed with @kylape on Slack - this is ok to close & we'll pass flags as needed in pod specs.

**Pull request closed without merging**